### PR TITLE
Fix gpfdist writable external table false success.

### PR DESCRIPTION
**Pull request by zhaorui20005**
When the segments write data to gpfdist through http protocol, the
socket may be reset because of the network failure. When gpfdist
detects this error, it tries to send an internal error code to the
clients side through the reset socket. But it will fail to inform
the clients because the socket is already unusable. The client side
usually gets an error when calling curl_easy_perform. So it will
retry the curl_easy_perform, and libcurl will try to connect to the
gpfdist will a new socket, and the new socket connection is regarded
as a new request by gpfdist. And at the time gpfdist accept this
socket, the fstream has already been closed by last error. So when
the new request tries session_attach, it will get an empty response
with no error. And so are all the following post requests sent by
segments. In this case, the download process(insert into writable
external table) will only write partail contents into the local file
and with no error msg.
So this is a false success. We should let gpfdist finish the process
soon after it detects a socket error.

For post requests, when gpfdist writes response back to the clients,
if an error is detected, gpfdists should set the session as an error.
So when the following request arrives, it can be rejected with
an internal error to let customer aware of an error happened before.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by kainwen — COMMENTED** [review]
I think we need a test case for this PR>

**Comment by zhaorui20005** [comment]
> I think we need a test case for this PR>

Sure, but this case relies on the network failure, and we have to make one of the socket which is created by libcurl to be reset while the download process. I didn't find a good way to do this.

File: src/bin/gpfdist/gpfdist.c [hunk 1]
```diff
@@ -1178,8 +1178,15 @@ static int local_send(request_t *r, const char* buf, int buflen)
...
 			/* close stream and release fd & flock on pipe file*/
-			if (r->session)
+			if (r->session && r->is_get)
 				session_end(r->session, 0);
+			/* For post requests, the error msg may not be transmited
+ 			 * to the client side because of network failure. So the 
+ 			 * session has to be set an error to inform the client
+ 			 * through the following request response with an 
+ 			 * internal error. */
+			else if (r->session && !r->is_get)
```
Comment (huiliang-liu) [inline src/bin/gpfdist/gpfdist.c:1188]: It is better to print a special log before `session_end()` to indicate this case.

Comment (zhaorui20005) [reply src/bin/gpfdist/gpfdist.c:1188]: There is one log at line 1181, and the log is unique in gpfdist.
And gpfdist is usually started for one purpose, so it is easy to identify if it is a post or get request, and there is also logs in session_end to identify if session_end has been called.
So I think the legacy logs are clear enough to locate this error.

**Review by huiliang-liu — APPROVED** [review]

**Review by lij55 — APPROVED** [review]

**Pull request merged**